### PR TITLE
Handle trait alias definitions

### PR DESCRIPTION
**Pull request by lowr**
Part of #2773 

This PR adds a bunch of structs and enum variants for trait aliases. Trait aliases should be handled as an independent item because they are semantically distinct from traits.

I basically started by adding `TraitAlias{Id, Loc}` to `hir_def::item_tree` and iterated adding necessary stuffs until compiler stopped complaining what's missing. Let me know if there's still anything I need to add.

I'm opening up this PR for early review and stuff. I'm planning to add tests for IDE functionalities in this PR, but not type-related support, for which I put FIXME notes.

### CODE REVIEW

**Comment by flodiebold** [comment]
Looks good to me :tada:

File: crates/hir/src/symbols.rs [hunk 1]
```diff
@@ -153,6 +153,10 @@ impl<'a> SymbolCollector<'a> {
                     self.push_decl(id, FileSymbolKind::Trait);
                     self.collect_from_trait(id);
                 }
+                ModuleDefId::TraitAliasId(id) => {
+                    // FIXME: Add `FileSymbolKind::TraitAlias`?
+                    self.push_decl(id, FileSymbolKind::Trait);
```
Comment (lowr) [inline crates/hir/src/symbols.rs:158]: I'm not really familiar with this area of code. Should I just add `FileSymbolKind::TraitAlias`?

File: crates/ide-completion/src/render.rs [hunk 2]
```diff
@@ -367,6 +367,8 @@ fn res_to_kind(resolution: ScopeDef) -> CompletionItemKind {
         ScopeDef::ModuleDef(Const(..)) => CompletionItemKind::SymbolKind(SymbolKind::Const),
         ScopeDef::ModuleDef(Static(..)) => CompletionItemKind::SymbolKind(SymbolKind::Static),
         ScopeDef::ModuleDef(Trait(..)) => CompletionItemKind::SymbolKind(SymbolKind::Trait),
+        // FIXME: Add `SymbolKind::TraitAlias`?
+        ScopeDef::ModuleDef(TraitAlias(..)) => CompletionItemKind::SymbolKind(SymbolKind::Trait),
```
Comment (lowr) [inline crates/ide-completion/src/render.rs:371]: Also here.

Comment (lowr) [inline crates/ide-completion/src/render.rs:371]: Also here

File: crates/hir/src/symbols.rs (hunk 1 above)
Comment (Veykril) [reply crates/hir/src/symbols.rs:158]: I think adding it for completeness sake should be good

**Comment by lowr** [comment]
Update: I began adding test, which quickly made me realize that I forgot to impl [`ToDef`](https://github.com/rust-lang/rust-analyzer/blob/27239fbb58a115915ffc1ce65ededc951eb00fd2/crates/hir/src/semantics.rs#L1535) - which wasn't straightforward because that trait pretty much requires one AST type to map to one HIR type whereas (previously) `ast::Trait` maps to either `hir::Trait` or `hir::TraitAlias`. There are 2 ways to solve:

- map `ast::Trait` to `Either<hir::Trait, hir::TraitAlias>`
- split `ast::TraitAlias` from `ast::Trait`

I've implemented the second approach as it seemed simpler, but I'm happy to hear opinions and thoughts.

**Comment by lowr** [comment]
This is now ready for full review. The commits are starting to get cluttered, I can squash relevant commits after review if desired.

Fully working IDE functionalities:
- doc_links
- file_structure
- move_item
- navigation_target
- runnables
- highlight_related

Partially working IDE functionalities (generally working for trait alias itself, but needs type support for full functionalities):
- goto_definition
- goto_type_definition
- hover
- inlay_hints
- moniker
- references
- rename
- signature_help

non-applicable functionalities: all others

**Review by Veykril — APPROVED** [review]

**Comment by lowr** [comment]
Rebased and organized commits. There's no diff in the final tree modulo now resolved conflict (which was merely one import statement). This PR should be good to go!

**Comment by lowr** [comment]
Might be worth noting explicitly in the changelog that we haven't implemented trait alias in our type system.

FYI, I was going to implement type support for trait alias, but after seeing rust-lang/rust#107614, I'm putting it on hold until types/lang teams decide what to do with it.

**Comment by HKalbasi** [comment]
@lowr Do you know what is the current state of trait aliases? Did the dust settle?

**Comment by lowr** [comment]
@HKalbasi I'm not really sure about its current state. The PR I mentioned in my last comment has been merged but [without any decisions][c] involved (which is reasonable as it's an unstable feature after all).

[c]: https://github.com/rust-lang/rust/pull/107614#issuecomment-1499830959

**Pull request merged**